### PR TITLE
CSHARP-4350: Add AWS Lambda tests

### PR DESCRIPTION
**Pull request by adelinowona**
This is the C# implementation for automated testing of a deployed lambda function.

- Added the lambda function in tests/LambdaTests
- Created the evergreen tasks to execute the scripts in drivers-evergreen-tools. (https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh)

Spec for this changes can be found here: [mongodb/specifications@master/source/faas-automated-testing/faas-automated-testing.rst](https://github.com/mongodb/specifications/blob/master/source/faas-automated-testing/faas-automated-testing.rst?rgh-link-date=2023-02-14T20%3A45%3A23Z)

What is the motivation for this change?
- WRITING-11435/[DRIVERS-2384](https://jira.mongodb.org/browse/DRIVERS-2384)
- The scripts in the evergreen tasks automate running everything described in the drivers ticket above and produces output we can inspect.

### CODE REVIEW

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
+using System;
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:1]: Do we need a copyright?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                    .Subscribe<CommandFailedEvent>(Handle)
+                    .Subscribe<ConnectionCreatedEvent>(Handle)
+                    .Subscribe<ConnectionClosedEvent>(Handle);
+            };
+
+            // Client is used for all requests
+            _mongoClient = new MongoClient(clientSettings);
+        }
+
+        public async Task<APIGatewayProxyResponse> LambdaFunctionHandler(APIGatewayProxyRequest apigProxyEvent, ILambdaContext context)
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:54]: async method names usually end in `Async`, but maybe this name is forced on us externally?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            var col = _mongoClient.GetDatabase("lambdaTest").GetCollection<BsonDocument>("test");
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:56]: We avoid abbreviations. Use `collection`?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            var testDoc = new BsonDocument("n", 1);
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:57]: Use `testDocument`?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        // actions for monitored events
+        private void Handle(ServerHeartbeatSucceededEvent @event)
+        {
+            _totalHeartbeatCount++;
+            _totalHeartbeatDurationMs += @event.Duration.TotalMilliseconds;
+            if (@event.Awaited && !_serverHeartbeatWasAwaited)
+            {
+                _serverHeartbeatWasAwaited = true;
+            }
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:100]: Could potentially be simplified to:

```_serverHeartbeatWasAwaited |= @event.Awaited;```

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        private void Handle(ServerHeartbeatFailedEvent @event)
+        {
+            _totalHeartbeatCount++;
+            _totalHeartbeatDurationMs += @event.Duration.TotalMilliseconds;
+            if (@event.Awaited && !_serverHeartbeatWasAwaited)
+            {
+                _serverHeartbeatWasAwaited = true;
+            }
```
Comment (rstam) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:110]: Could potentially be simplified to:

```_serverHeartbeatWasAwaited |= @event.Awaited;```

**Review by rstam — CHANGES_REQUESTED** [review]
I only reviewed the C# code because I am unfamiliar with the rest of the Evergreen and other configuration.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 1 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:1]: Yep. Done

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 2 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:54]: Added `Async`, the function name isn't forced on us externally.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 3 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:56]: Done.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 4 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:57]: Done.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 5 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:100]: Done.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 6 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:110]: Done.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 7]
```diff
@@ -0,0 +1,152 @@
+/* Copyright 2010-present MongoDB Inc.
```
Comment (BorisDog) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:1]: Looks like we are going to have other FaaS tests, Do we want add `FaaSTests` parent folder?

File: tests/LambdaTests/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# AWS Lambda Testing
+
+This project contains source code and supporting files for a serverless application that you can deploy with the SAM CLI. It includes the following files and folders.
+
+- src - Code for the application's Lambda function.
```
Comment (BorisDog) [inline tests/LambdaTests/README.md:5]: `src` is probably not the right folder.

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            }
+            else
+            {
+                responseBody = new Dictionary<string, string>
+                {
+                    { "averageCommandDurationMS", $"{_totalCommandDurationMs / _totalCommandCount :F3}"},
+                    { "averageServerHeartbeatDurationMS", $"{_totalHeartbeatDurationMs / _totalHeartbeatCount :F3}"},
+                    { "openConnections", $"{_openConnections}" },
+                    { "heartBeatCount", $"{_totalHeartbeatCount}" },
+                    { "CommandCount", $"{_totalCommandCount}" },
```
Comment (BorisDog) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:93]: lower case?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        private void Handle(ConnectionCreatedEvent @event)
+        {
+            _openConnections++;
+        }
+
+        private void Handle(ConnectionClosedEvent @event)
+        {
+            _openConnections--;
```
Comment (BorisDog) [inline tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:141]: This might be concurrent execution, so better to use `Interlocked.Dec`.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 7 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:1]: Done.

File: tests/LambdaTests/README.md (hunk 8 above)
Comment (adelinowona) [reply tests/LambdaTests/README.md:5]: Fixed!

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 9 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:93]: fixed!

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 10 above)
Comment (adelinowona) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:141]: Done!

File: tests/FaasTests/LambdaTests/README.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+- AWS SAM CLI
+- Docker daemon running with mongodb instance
+
+Build the application with the `sam build` command from the `tests/LambdaTests` folder.
+
+```bash
+sam build
+```
+
+The SAM CLI installs dependencies defined in `tests/LambdaTests/src/MongoDB.Driver.LambdaTest.csproj`, creates a deployment package, and saves it in a `.aws-sam/build` folder.
```
Comment (BorisDog) [inline tests/FaasTests/LambdaTests/README.md:23]: Update the path, or maybe make it relative?

File: tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs (hunk 10 above)
Comment (BorisDog) [reply tests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs]: Sorry I meant that we need to access `_openConnections` in a threadsafe manner, so need to increment it with Interlocked as well.

File: tests/FaasTests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            _openConnections++;
+        }
+
+        private void Handle(ConnectionClosedEvent @event)
+        {
+            Interlocked.Decrement(ref _openConnections);
+        }
+
+        private void Reset()
```
Comment (BorisDog) [inline tests/FaasTests/LambdaTests/MongoDB.Driver.LambdaTest/LambdaFunction.cs:145]: So why do we need reset?

**Review by BorisDog — CHANGES_REQUESTED** [review]
minor comments

**Review by rstam — APPROVED** [review]
LGTM

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**